Livecheck Formula DSL

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Addresses #7027, aims to add a Formula DSL for `livecheck`.

Just starting to implement this and I have a few doubts. Where would the functions for `skip`, `using`, etc. mentioned in the example be defined? Or would they be implemented in some other way?

So far (locally) I have tried this, below the definition of `livecheck(&block)`:
```ruby
def skip(reason)
  puts "Skipping: #{reason}"
end
```
However, I'm not sure if that's the way I should be doing it.

MikeMcQuaid commented:
> Just starting to implement this and I have a few doubts. Where would the functions for `skip`, `using`, etc. mentioned in the example be defined?

See e.g. `SoftwareSpec#patch` and `Patch.create` for how patches are done.

samford commented:
`skip` is one of the features that I've implemented, so you may also want to look at how it's handled in livecheck as part of the current `Hash` argument setup: https://github.com/Homebrew/homebrew-livecheck/blob/master/cmd/livecheck.rb#L104-L117

As you can see, the output happens in livecheck but we need to be able to be able to check for `skip` having been provided in the livecheck arguments (the block, in this context) and to be able to access the provided message (if any).

For what it's worth, I've been working on implementing the other features as time permits, so we will just need to update them to work with the livecheck DSL when the time comes (rather than implement them all from scratch). I'll try to keep you updated as features are added.

nandahkrishna commented:
Got it @samford. I'll look at `skip` for now, and keep track of changes in livecheck so we can add them here when they're done. Thanks.

nandahkrishna commented:
Just an update, working on understanding software_spec and checking for the presence of "skip" in the livecheck block. I'll try and make some changes by end of day.

vidusheeamoli commented:
> I traced the output of brew livecheck some-formula --debug and I see that any changes I make to formula.rb aren't reflected, whereas code in the new livecheck block is executed followed by livecheck's original code (usually throws an error). I'm checking this by editing some-formula.rb from homebrew/core.

Ran into a similar problem recently. I was making changes to my fork of `homebrew/core` instead of `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core`. 

Just do `cd $(brew --repo homebrew/core)` and make your changes there and they will be reflected right away! (https://github.com/Homebrew/brew/issues/6975#issuecomment-592954256)

Hope this helps. 😄

nandahkrishna commented:
Ah, I should've been more specific. I've been making changes in `/usr/local/Homebrew/Library/Homebrew/formula.rb` (using `cd $(brew --repo)`) and to some formula in `/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core`, which is the right place I believe. I'll provide a minimum working example here for more details. Thanks for helping @vidusheeamoli!

**some-formula.rb in homebrew/core:**
```ruby
...
  livecheck do
    puts "Not maintained" # For example!
  end
...
```

**formula.rb in Homebrew:** changes made in [this commit](https://github.com/Homebrew/brew/pull/7179/commits/f8573982b1b476097bbbf2810d1ec83a6ae15244)

I have the `livecheck` tap installed, and I execute `brew livecheck some-formula --debug`, which results in:
```
...
Not maintained
Error: some-formula: Version value must be a string; got a NilClass ()
```
However, doing this works (thanks to @samford's recent PR on adding `:skip` in livecheck)
```ruby
...
livecheck :skip => "Not maintained"
...
```

I realise this is because I've not really called my livecheck function (from the Formula DSL) anywhere, but was just trying to understand how the whole thing works. If I'm not wrong, `brew` takes care of choosing which Tap to execute and when I run `brew livecheck`, the tap's code is executed (giving a `command not found` error when the tap is not present). Surprisingly to me, the code from the livecheck block in the Formula is also executed (I'm working on understanding this).

So my question really is, how do I handle the DSL specified in the livecheck block of a formula using the function defined in `formula.rb`, and more importantly, how do I see if it works? Would it require changes in livecheck or an additional command here?

Hope I was clear, sorry for the confusion 😅

MikeMcQuaid commented:
> Would it be alright to use a gem such as sourcify and check for the presence of the `skip` directive in the Livecheck block? It didn't seem too elegant to me, let me know if you have a better solution.

Nope, sorry! This should be doable using our existing infrastructure.

> A couple more questions to understand how this works, if you don't mind 😅. I'm new to Ruby and hence a little slow when it comes to these things.

Happy to answer! Don't need to caveat; we're all learning here ❤️ 

> So my question really is, how do I handle the DSL specified in the livecheck block of a formula using the function defined in `formula.rb`, and more importantly, how do I see if it works? Would it require changes in livecheck or an additional command here?

I'm not sure I totally understand you but will do my best to answer anyway!

It's up to you whether you match or deviate from the existing `livecheck` syntax in the Homebrew/homebrew-livecheck tap's `brew livecheck` command. What you want to be doing is making changes in three places:

1. the Homebrew/brew repository (this one!) to add the `livecheck`  DSL to `formula.rb`
2. the Homebrew/homebrew-core repository to add `livecheck` to various files in `Formula`
3. the Homebrew/homebrew-livecheck tap's `brew livecheck` command to read from `Formula#livecheck`/`Formula.livecheck` instead of the `Livecheckables` directory

You'll probably need to have all three PRs open at once for your local testing and we'll merge 1. first (which will cause no disruption and need no changes), then 2. for e.g. a single formula, then 3. to use the `livecheck` DSL if it's available or  fall back to `Livecheckables`. When all `Livecheckables` have been moved to Homebrew/homebrew-core the fallback code in `brew livecheck` can be removed (and we may even consider migrating the command to Homebrew/brew.

Does that help/make sense?

nandahkrishna commented:
Alright, @MikeMcQuaid. Thanks, that did help!

MikeMcQuaid commented:
Still in progress, stalebot.

nandahkrishna commented:
Hey, so sorry for the long gap. Just made a minor modification for the `skip` feature in livecheck. I'm not sure my style is right and I hope to improve this based on your suggestions. I sort of tested this implementation with a smaller program, do let me know if there are changes to be made.

Also spent time reading the updates discussed in the livecheck DSL issue. Didn't understand everything but I think I've got a fair idea. I'll get working on any updates as soon as possible!

nandahkrishna commented:
@MikeMcQuaid, I don't know if I'm missing something but I'm not able to get the Formula's name within the DSL - I've tried using `#{full_name}` or `#{name}` - they either throw an error or give me the entire class name. What should I be using instead?

MikeMcQuaid commented:
@nandahkrishna When you say "within the DSL": where do you mean specifically? Thanks!

nandahkrishna commented:
Ah, I meant within the `skip` function:
```ruby
def skip(skip_msg)
  skip_msg = if skip_msg.is_a?(String) && !skip_msg.empty?
    " - #{skip_msg}"
  else
    ""
  end
  puts "#{Tty.red}#{name}#{Tty.reset}: skipped#{skip_msg}"  # right here
end
```

MikeMcQuaid commented:
> they either throw an error or give me the entire class name.

Try `self.name`.

nandahkrishna commented:
I did try that, earlier, here's an example of what happens (it's somehow similar to using just `name`):
```
Formulary::FormulaNamespace6b9a2f4a9f1d162261b2460cb09db23e::SomeFormula: skipped - Not maintained
```

`brew style` says that this is an offense:
```ruby
Offenses:

formula.rb:2638:25: C: Redundant self detected.
      puts "#{Tty.red}#{self.name}#{Tty.reset}: skipped#{skip_msg}"
                        ^^^^^^^^^

769 files inspected, 1 offense detected
```
Alternatives I'm looking at:
* A function to get the name, perhaps
* Modifying the Core Formula's `some-formula.rb` to include `some-formula` as an argument

nandahkrishna commented:
So I managed to handle it this way, @MikeMcQuaid - as you'd suggested earlier, I made a small modification in homebrew/livecheck's code which checks if "skipped" is part of the string returned from the execution of a Formula's `livecheck` block:

**homebrew/livecheck's cmd/livecheck.rb**
```ruby
if latest.to_s.include? "skipped"
  puts "#{Tty.red}#{formula_name(formula)}#{Tty.reset}: #{latest.to_s}"
  return
end
```

To elucidate, livecheck already determines whether the arguments to it are from the block (as a `Proc`) or from the Livecheckable. The latest version is determined based on these arguments. However the presence of `skip` in the block wasn't handled, so I wrote that code to handle it.

Now if a Formula has a `livecheck` block, `brew livecheck some-formula` will correctly display the formula's name (within livecheck's code) and the skip message.

Do let me know if you have any suggestions on this. I'll be opening up a PR in homebrew/livecheck for this fallback code as well. Thanks!

samford commented:
I'm in the process of reviewing the current `skip` implementation and I'll be providing some feedback on how it (as well as the PR in livecheck) could be changed to be a bit more reliable and to keep some of this logic in livecheck (where it's more appropriate) instead of in `skip`.

Just wanted to give everyone a heads up in the interim time (should be up within an hour or two at most), so we don't end up going in two different directions.

samford commented on Library/Homebrew/formula.rb:
```diff
@@ -2620,6 +2623,23 @@ def test(&block)
+
+    # Handles the `skip` feature for the livecheck Formula DSL
+    # WIP
+    def skip(skip_msg)
+      skip_msg = if skip_msg.is_a?(String) && !skip_msg.empty?
+        " - #{skip_msg}"
+      else
+        ""
+      end
+      "skipped#{skip_msg}"
+    end
```
```suggestion
    def skip(skip_msg = nil)
      @skipped = true
      @skip_msg = skip_msg
    end
```

I think it would be simpler and more in keeping with other methods in `formula.rb` if `skip` simply set a `@skipped` instance variable to `true` (indicating that `skip` was present in the `livecheck` block in the formula) and set a `@skip_msg` instance variable to the provided message or `nil` by default.

I think it's more appropriate to handle the output for the "skipped" text and printing `skip_msg` in  Homebrew/livecheck instead. The trouble is that we can only do that if we make `@skipped` and `@skip_msg` available to livecheck through public methods (e.g., `skipped?` and `skip_msg`). From my testing, this would seemingly also require adding `skipped?` and `skip_msg` methods to `extend/formula.rb` in Homebrew/livecheck (which simply call the class methods).

This bumps into the same issue I had before while working on the version-filtering feature, where livecheck's access is limited and we seemingly have to expose things that we may not want to expose in `formula.rb`. If it's not appropriate to expose `skipped?` and `skip_msg` methods, we would seemingly have to wait until after Homebrew/livecheck is part of Homebrew/brew to do add implementation for `skip` there.

nandahkrishna replied:
Thanks for the suggestion, @samford. I did run into the issue of calling/using private methods when I was trying out different implementations. For now, I guess, I'll go ahead and make this change.

MikeMcQuaid commented:
> we would seemingly have to wait until after Homebrew/livecheck is part of Homebrew/brew to be able to properly handle `skip`.

I would make this a priority over doing too much more work in Homebrew/livecheck, @nandahkrishna.

nandahkrishna commented:
Alright @MikeMcQuaid.

> I would make this a priority...

Just to clarify, you'd want me to work on transitioning livecheck into Homebrew/brew, or just the other DSL features?

MikeMcQuaid commented:
> Just to clarify, you'd want me to work on transitioning livecheck into Homebrew/brew right?

Yup!



> And if you don't mind, could you just point me to the changes I'd have to make, for example, including command support and adding livecheck's functionality to, say, `formula.rb` or `software_spec.rb`?

`formula.rb` and a new `livecheck.rb` to define a `LiveCheck` class to `class_eval` the block passed to `Formula.livecheck`.

nandahkrishna commented:
A doubt on the implementation of `skip` in the `LiveCheck` class - would it still just be setting the attributes `@skipped` and `@skip_msg`, or would it be a proper skip implementation as is available in livecheck, since we're planning a migration of livecheck into Homebrew/brew? Thanks.

nandahkrishna commented:
I've made some changes, with the flow being as follows:

* The `livecheck` block specified in the Formula is `instance_eval`ed by `Formula.livecheck`
* If `skip` is given, `Formula.skip` by means of `class_eval` sets the `@skipped` and `@skip_msg` attributes of `LiveCheck` to the appropriate values.

If I've misunderstood and this isn't the right way, let me know!

samford commented:
I've tinkered with this a bit and I believe I understand what the intention of the `Livecheck` class may be here but @MikeMcQuaid will have to confirm.

One of the issues that we face in implementing the livecheck DSL is that `Formula` already uses `url` to refer to the stable URL, so if `livecheck` set a `@url` instance variable it would clash with the existing use of `url`.

We have to deal with that naming collision somehow and it may make sense to handle this with a class that we can use to both store instance variables as well as implement the methods that are executed in the `livecheck` block (e.g., `url`, `regex`, `skip`, etc.). I assume this is the purpose of the `Livecheck` class here.

We'll need to make this object accessible somehow, so we can access it (and its instance variables) in livecheck. However, `Formula.livecheck` would seemingly call `Formula#livecheck`, so we may have to name it something else like `Livecheckable` (so we could surface it like `formula.livecheckable.url`, `formula.livecheckable.skip?`, etc.). [The `livecheck` block in formulae is fulfilling the same purpose as the livecheckables in Homebrew/homebrew-livecheck, so the "livecheckable" name makes sense to me in this context.]

In this scenario, `Formula#livecheck` could look like this:

```ruby
    def livecheck(&block)
      @livecheckable = Livecheckable.new
      @livecheckable.instance_eval(&block)
    end
```

With this setup, the `skip` method would move to the `Livecheckable` class and you would remove the `class_eval` part there (simply setting the instance variables). You can then implement a `skip?` method that simply returns `@skip` (i.e., we would use `formula.livecheckable.skip?` in livecheck). `@skip_msg` could simply be surfaced with `attr_reader`.

What do you think about all this (the class setup, naming, etc.), @MikeMcQuaid? I've tested this out and it works but the only issue I'm having is that I can't seem to make `@livecheckable` accessible for a `Formula` object when testing this in `brew irb`. [I've tried using `attr_reader :livecheckable` as well as creating a `livecheckable` method that simply returns `@livecheckable` but both setups return `nil` even though `Livecheckable#initialize` and `Livecheckable#skip` are being executed, so I'm missing something.]

Just to be clear @nandahkrishna, please don't implement what I've outlined here until others have given their thoughts and we've agreed on how to proceed.  These are just my own ideas on this and should be taken with a grain of salt.

---

Something else that I've been meaning to bring up is how the parts of livecheck that aren't livecheckables would be incorporated into Homebrew/brew. Currently the folder structure in Homebrew/homebrew-livecheck looks like this (ignoring the README and old CHANGELOG):

* cmd
  * livecheck.rb (implements the `brew livecheck` command)
* livecheck
  * commands
    * cat.rb (doesn't appear to be used anymore, so we may be able to simply remove it)
  * extend
    * formula.rb (extensions to `Formula` related to livecheck)
    * formulary.rb (extensions to `Formulary` related to livecheck)
  * heuristic.rb (implements `version_heuristic()`, which checks URLs for versions and returns the newest)
  * utils.rb (various utility functions, only used in `heuristic.rb` currently)

Outside of `cat.rb` (which I'll probably remove in a PR), the other files are all necessary for `brew livecheck` to work and they would need to be integrated into Homebrew/brew as well. I can't provide any insight on how to accomplish this but I wanted to bring the issue up, so someone else can address it.

samford commented:
Just as a follow-up to my previous comment after testing this further: the aforementioned class setup worked for `skip` because there was no naming collision between the method (`#skip`) and the instance variables (`@skipped`, `@skip_msg`).

In testing out `url`, I ran into an issue where `Livecheckable#url`, which implements the DSL method, prevents us from being able to surface the `@url` instance variable at `formula.livecheckable.url` (since that's `Livecheckable#url`). `instance_variable_get(:@url)` works but `formula.livecheckable.instance_variable_get(:@url)` is far less ergonomic than `formula.livecheckable.url`.

Alternatively, we could store the data from the livecheck block in an `args` hash (instead of instance variables), leading to `formula.livecheckable.args[:url]`, which is relatively close to the current `formula.livecheck_args[:url]`. A different name may make more sense, though (e.g., `values` rather than `args`?).

Does anyone have any ideas for a better way of dealing with this particular issue (or this setup in general)?

MikeMcQuaid commented:
> In this scenario, `Formula#livecheck` could look something like this:

Indeed, that's what we should do. See `def stable(&block)` for an example of that in action for the `SoftwareSpec` class.

samford commented:
I went down this rabbit hole further and created a working implementation where `formula.livecheck.url` will return the URL (or `nil`, if `url` isn't provided in the `livecheck` block). The basic idea is this:

* Keep the methods for the `livecheck` DSL in an inner class in the `Livecheck` class (the `class << self` block). These methods will set instance variables like `@url`, `@regex`, `@skip`, `@skip_msg`, etc. These instance variables belong to the inner class, not the outer `Livecheck` class.
* Create getter methods in the `Livecheck` class, which use `self.class.instance_variable_get()` to return an instance variable from the inner class. These are the methods that allow us to get a value like `formula.livecheck.url`.
  * It's necessary to use `instance_variable_get()` because `self.class.url` would call the `#url` method instead of returning the inner `@url` instance method.
  * We want to follow the typical Ruby convention of having a question mark at the end of a method that returns a boolean, so we'll want to have a `#skip?` getter method here instead of `#skip`.
* Modify the `#livecheck` method in `Formula`'s inner class to do `@livecheck.class.instance_eval(&block)` (instead of `@livecheck.instance_eval(&block)`, as mentioned above), so the method calls in the `livecheck` block use the methods in the inner class of `Livecheck` (not the getter methods in the main class body).
* Modify the empty `#livecheck` method in `Formula`'s main class (not the inner class) to return the `@livecheck` instance variable from the inner class using `self.class.instance_variable_get(:@livecheck)`. This gives us access to the `@livecheck` instance variable at `formula.livecheck`.

I've confirmed that this setup works as expected in Homebrew/homebrew-livecheck with some simple modifications. If we wanted this to work in livecheck before it's merged into Homebrew/brew, we would also have to incorporate the `Formula` changes into livecheck's `Formula` extension and modify `_latest` to properly work with the DSL. I've made all the required changes in a local branch and have `url`, `regex`, and `skip` working with a `livecheck` block in a formula (in addition to existing support for livecheckables), so this works for all existing features.

What do you think about this, @MikeMcQuaid? If you need to see my implementation to understand this, I can provide you with the code privately somehow but I wanted to avoid publishing the code publicly (to allow Nanda to have the experience of doing the implementation).

MikeMcQuaid commented:
> * which use `self.class.instance_variable_get()` to return an instance variable from the inner class

We should avoid this; instance variables are an implementation detail that shouldn't be accessed outside their class.



> * instead of `@livecheck.instance_eval(&block)`, as mentioned above

I think an `instance_eval` would be preferable to allow easier testing.



> What do you think about this, @MikeMcQuaid? If you need to see my implementation to understand this, I can provide you with the code privately somehow but I wanted to avoid publishing the code publicly (to allow Nanda to have the experience of doing the implementation).

I think I'll need to see implementation to understand. That said, I think it would be best to provide guidance to @nandahkrishna without stating how the problem should be solved or how exactly how code should be written (before it is).

samford commented:
Seeing the implementation may ease your concerns above (regarding `instance_variable_get()` and `@livecheck.class.instance_eval(&block)`), so I'll send you something over Slack soon.

> I think it would be best to provide guidance to @nandahkrishna without stating how the problem should be solved or how exactly how code should be written (before it is).

I agree. I was trying to strike a balance (in some way) but I ended up erring on the side of providing detail so that the foundation of the DSL ends up solid (as it impacts how we access the DSL information in livecheck and it should be ergonomic). I'm particularly invested in seeing this turn out well and I was perhaps overzealous about sharing my discoveries. In hindsight, that wasn't the best way to go about things and it would have been better to just discuss some of these implementation details with you privately.

Going forward, I'll try to leave things more open-ended when providing guidance.

MikeMcQuaid commented:
> Going forward, I'll try to leave things more open-ended when providing guidance.

Thanks ❤️ 



> as it impacts how we access the DSL information in livecheck and it should be ergonomic

Just to clarify for @nandahkrishna: it may be that `brew livecheck` should or could end up better as a command in Homebrew/brew from this so I wouldn't worry too much about the existing `brew livecheck` implementation but instead figure out how best to implement the DSL in Homebrew/brew so that it can be used in Homebrew/homebrew-core and we can move on from there.

nandahkrishna commented:
Alright @MikeMcQuaid, I'll prioritise making the DSL as effective and efficient as possible.

nandahkrishna commented:
I've made some modifications:
* `skip` has now been moved to the `Livecheck` class
* Instance variable values are set using the functions on `instance_eval`
* A `to_hash` function returns a `Hash` of all instance variables
* The `livecheck` function in the main`Formula` class returns this `Hash` - as we wanted to be able to access the `@livecheck` instance's variable values
I've tried to test it and it seems to work, but I'm not sure I'm testing this the right way. Any suggestion on how I should be testing the DSL?
@MikeMcQuaid @samford

samford commented on Library/Homebrew/livecheck.rb:
```diff
@@ -0,0 +1,22 @@
+    @skip_msg = skip_msg
+  end
+
+  def url(val = nil)
```
This is just my own perspective but it may be better for us to allow this to error if `url` is in a `livecheck` block and no argument is provided. That is to say, a default value doesn't seem appropriate here, as a passed URL should be required.

What does everyone else think about this?

MikeMcQuaid replied:
The default value is used in these sort of DSLs to allow the use of both `livecheck.url` and `livecheck do; url "$URL; end` i.e. this method is both a getter and a setter.

nandahkrishna replied:
That would mean it's best to remove the `ArgumentError` I've raised when the value is blank, so that it can function as both a getter and a setter?

MikeMcQuaid replied:
> That would mean it's best to remove the `ArgumentError` I've raised when the value is blank, so that it can function as both a getter and a setter?

Yup 👍

nandahkrishna commented:
> Could you explain specifically how you tested this?

Ah, I was testing it within Homebrew/brew (without any changes to livecheck, as Mike had mentioned earlier). I was using a setup quite similar to yours - but with some changes here and there to circumvent `LoadError`s as I didn't know of `brew irb`. I modified `anime-downloader` (no reason, just that I initially contributed it 😅) to include a `livecheck` block. Thanks for the suggestion on `brew irb`!

Using the `--debug` flag earlier in livecheck was helpful in understanding the sequence of things, so I did manage to get some of it right, like using `Formulary` to load and then call `formula.livecheck`.

> Simple debug statements like these can sometimes help you understand what's going on.

Sure thing! No hesitation to add one after every line so far 😬

I'll make the changes you've suggested and test it with this new setup. Thanks.

nandahkrishna commented on Library/Homebrew/formula.rb:
```diff
@@ -1813,6 +1814,11 @@ def test_defined?
   def test; end
 
+  # @private
+  def livecheck
+    @livecheck.to_hash
+  end
```
Ah, yes I do understand. I wasn't testing it the right way, maybe that's why I didn't realise earlier. I'll work on it!

nandahkrishna commented:
> Could you explain specifically how you tested this?

Ah, I was testing it within Homebrew/brew (without any changes to livecheck, as Mike had mentioned earlier). I was using a setup that involved writing another `.rb` file to load the `Formula` and call the function. I modified `anime-downloader` (no reason, just that I initially contributed it 😅) to include a `livecheck` block. Thanks for the suggestion on `brew irb`!

Using the `--debug` flag earlier in livecheck was helpful in understanding the sequence of things, so I did manage to get some of it right, like using `Formulary` to load. However I was then calling `formula.class.instance_variable_get(:@livecheck)`. `formula.livecheck`, as you said, did not work as I'd expected.

> Simple debug statements like these can sometimes help you understand what's going on.

Sure thing! No hesitation to add one after every line so far 😬

I'll make the changes you've suggested and test it with this new setup. Thanks.

nandahkrishna commented:
* Inspired by how `desc` and `homepage` are accessible through an instance of the loaded formula, I removed the method `Formula.livecheck` and have instead used `delegate`, so that the inner class' livecheck object is accessible. I did try a couple other things, such as making the method return `self.class.instance_variable_get(:@livecheck)` but this seemed more apt. Let me know if this is acceptable.
* I've removed the `= nil` default argument for `url`, as it doesn't make sense to leave it empty if specified in the first place.
* I re-wrote the inner class' `livecheck` method in a manner similar to `stable`.

A question: We would need to add audits to ensure the DSL specification is proper, yes? Just asking for now so that I can start reading up and thinking about it.

MikeMcQuaid commented:
> To start with, you'll need to add a `livecheck` block to a formula
> I've mostly been using `brew irb` to test this for the time being

This is what I'd recommend 👍 

> To load a formula, I've been using `f = Formulary.factory('zzz')`

This works but to save keystrokes you can do `f = :zzz.f`

> We would need to add audits to ensure the DSL specification is proper, yes?

Maybe! We only need audits when there are values that aren't desirable but are valid input. At this stage, you're probably better to error out in the functions on bad input instead.

nandahkrishna commented:
> This works but to save keystrokes you can do `f = :zzz.f`

Wow, thanks!

> you're probably better to error out in the functions on bad input

Great, I'll do just that.

Updated to add support for `regex` as well, and some better handling for `skip`.

samford commented on Library/Homebrew/livecheck.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def skip(skip_msg = nil)
+    @skip = true
+    @skip_msg = if skip_msg.blank?
+      nil
+    else
+      skip_msg
+    end
```
```suggestion
    @skip_msg = skip_msg.blank? ? nil : skip_msg
```

Since this is so simple (conditionally choosing the value to set), you can use the ternary operator (`?:`) to make this less verbose.

MikeMcQuaid commented on Library/Homebrew/livecheck.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def to_hash
+    Hash[instance_variables.map { |name| [name, instance_variable_get(name)] }]
```
I think it would be better to manually specify this hash; it's easier to read and figure out what the format is.

nandahkrishna commented:
I've added `#livecheckable?`, and `test/livecheck_spec.rb` with unit tests. Do let me know if there are changes to make.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -2186,6 +2198,9 @@ def method_added(method)
     attr_rw :homepage
 
+    # Whether a Livecheck specification is defined or not.
+    attr_reader :livecheckable
```
Think this should be a `livecheckable?` method too.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -353,6 +354,17 @@ def bottle
+  # Is a Livecheck specification defined for the software?
+  # @private
+  def livecheckable?
+    self.class.livecheckable == true
```
I'd delegate this to `livecheckable?`

MikeMcQuaid commented on Library/Homebrew/livecheck.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def skip(skip_msg = nil)
+    @skip = true
+    @skip_msg = skip_msg.blank? ? nil : skip_msg
```
```suggestion
    @skip_msg = skip_msg.presence
```

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -2628,6 +2643,16 @@ def test(&block)
     end
 
+    # Livecheck can be used to check if there are any updates to the Formula.
+    # WIP
```
Would be good to have comments on all these methods (and those in `livecheck.rb`)

nandahkrishna replied:
Sure, I'll get working on them. Just trying to frame them in such a way that they're clear and concise. Looking forward to your suggestions once they're done.

MikeMcQuaid replied:
Sure thing 👍. They don't need to be more complex than typical methods in `formula.rb`. Think of those in `formula.rb` from the perspective of a contributor trying to write a new formula and in `livecheck.rb` from a Homebrew maintainer wanting to refactor them.

MikeMcQuaid commented on Library/Homebrew/test/livecheck_spec.rb:
```diff
@@ -0,0 +1,63 @@
+      expect(livecheckable.instance_variable_get(:@skip_msg)).to be nil
+    end
+
+    it "sets the instance variable skip to true and skip_msg to nil when the argument is blank" do
```
Can skip this test as it's essentially just testing the behaviour of `blank?` (and isn't needed if you use `.presence`)

nandahkrishna commented:
Should I add the livecheck data to Formula's to_hash method? From what I understand, it returns as full representation of the formula as a Hash (containing all details) - so should livecheck be included there as well?

MikeMcQuaid commented:
> Should I add the livecheck data to Formula's to_hash method?

I think: _maybe_ eventually but definitely not yet. Thanks for asking, though!

nandahkrishna commented:
Added comments and made the changes you'd suggested, @MikeMcQuaid. Now as for testing the DSL from a Formula, would it be in `formula_spec.rb`, similar to the test for `spec integration`?

MikeMcQuaid commented:
Looking good! Once CI is 🟢 I'm happy to ✅ this.

nandahkrishna commented:
Fixed these issues:
* Enclosed Regexp within `()` to avoid ambiguity style warning
* Excluded `formula_spec.rb` from the Block Length cop in `rubocop.yml` as adding tests made block size go beyond 1100 limit (1115).

nandahkrishna commented:
CI is 🟢! Waiting for any other changes before we can merge this, thanks!

MikeMcQuaid commented:
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @nandahkrishna!
